### ISSUE TITLE
func without currying?

### ISSUE THREAD

**Issue by tmcw** [gcanti/tcomb#96]
I'm using tcomb to add strictness to actions in a Flux application. It's awesome!

I was surprised by the default behavior of function:

```js
var tcomb = require('tcomb');
var add = tcomb.func([tcomb.Num], tcomb.Num).of(function(a) { return a + 1; })
add();
```

I expected this to throw, but really it curries `func` by default, so calling functions without arguments just returns the functions. I'd much rather have an API that throws on missing functions than silently curries, as this is doing.

I can see from the earlier discussion https://github.com/gcanti/tcomb/issues/84 that this is a tricky problem. Could currying be optional for the `func` method?

**Comment by gcanti** [comment]
cc @beckyconning

**Comment by beckyconning** [comment]
i don't think that currying should be optional as its easy to have a function from a tuple to something else. e.g.

```
\\ f :: (Num, Num) -> Num
var f = T.func([T.tuple([T.Num, T.Num])], T.Num).of(function (xs) { return xs[0] + xs[1]; });
```

however i agree that functions called without arguments should throw type errors. e.g.

```
\\ f :: Num -> Num -> Num
var f = T.func([T.Num, T.Num], T.Num).of(function (x, y) { return x + y; });
f(); \\ was not called with a number should throw type error
```

**Comment by beckyconning** [comment]
Although you could wrap whatever you are expecting to take the result of that argument using the func combinator to ensure that it takes a `Num` rather than a `Func`.

```var safeRender = T.func([T.Num], U.Promise).of(render);```

**Comment by tmcw** [comment]
> i don't think that currying should be optional as its easy to have a function from a tuple to something else. 

I'm adding tcomb to an existing API: in order to use tuples, I'd need to change all method signatures and everywhere those methods are called, right?

**Comment by beckyconning** [comment]
i think it would probably be easier and more valuable to wrap the api methods which you are calling using the `func` combinator than to use the tuple approach.

**Comment by tmcw** [comment]
I'm not totally sure what that means: for instance, in my app I have a method like this:

```js
var actionSignOrder = func([Str, Obj], Prom).of(function(id, data) {
      return MapboxClient.signOrder(id, data)
        .then(orders => {
          dispatch({
            actionType: OrdersConstants.ORDER_SIGN
          });
      });
    })
```

This is a method with two required parameters. Somewhere else in my application, I had a call to this method that gave it only a single `str` argument. Before, that call would fail because MapboxClient would try to get a property of `data`, which would be undefined, and throw. Ideally, that call would fail immediately because it didn't include a required parameter, and tcomb could give one of its informative error messages immediately. But with automatic currying, invoking `actionSignOrder(str)` fails silently, which is arguably worse than either of the other alternatives.

**Comment by gcanti** [comment]
I'm adding an optional boolean param `uncurried` to `Func.of(f, uncurried)`

```js
var Type = func([Num, Num, Num], Num);
var sum = Type.of(function (a, b, c) {
  return a + b + c;
}, true); // <= uncurried = true
sum(1, 2, 3); // => 6
sum(1, 2); // throws
```

**Comment by gcanti** [comment]
I changed my mind about the next version of tcomb. I'll ship a v2.0 so I can do breaking changes. The uncurried version will be the default and I'll add a curried optional param.

**Comment by gcanti** [comment]
published in v2.0.0-beta
